### PR TITLE
Create a Sidebar Component

### PR DESCRIPTION
**Pull request by pauloptimizely**
## Summary

This a base Sidebar component that can be used for both sidebar navigation and pop-over drawers.

At the moment, the Sidebar is just a highly opinionated div. More sub-component will be added in the future.

**Possible Future component**
- SidebarHeader
- SidebarNavList
- SidebarNavListItems

## Test Plan

run the test suite

```
npm test
```

verify everything looks good in storybook

```
npx yarn storybook
```

## Jira

https://optimizely.atlassian.net/browse/MGMT-1258

### CODE REVIEW

**Comment by daverau-optimizely** [comment]
Hey @pauloptimizely, is this for the event debugger that Ron designed?

I was just wondering if we should include the `X` close button for dismissing this, or would you consider that separate from the component?

File: src/components/Sidebar/index.scss [hunk 1]
```diff
@@ -0,0 +1,24 @@
+.oui-sidebar {
+  background: $base-white;
+  display: flex;
+  flex-direction: column;
+  height: 100%;
+  position: absolute;
+  transition: all 0.3s;
+  top: spacer(0);
+  width: spacer(0);
```
Comment (daverau-optimizely) [inline src/components/Sidebar/index.scss]: We haven't used `spacer(0)` before — can we just use `0` here?

Also somewhat minor stuff but I'd be inclined to move `flex, flex dir, height, absolute position, and bg color` to helpers so: in the react render you can include these classes and remove the CSS rules: `flex flex--column height-1-1 position--absolute background--white` — I'm a stickler for not adding CSS to the global stylesheet unless it's absolutely necessary.

File: src/components/Sidebar/index.scss [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+  z-index: 99;
+
+  &--docked {
+    position: relative;
+    height: initial;
+  }
+
+  &--open-left {
+    left: spacer(0);
```
Comment (daverau-optimizely) [inline src/components/Sidebar/index.scss]: `0`

File: src/components/Sidebar/index.scss [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+  }
+
+  &--open-right {
+    right: spacer(0);
```
Comment (daverau-optimizely) [inline src/components/Sidebar/index.scss]: `0`

File: src/components/Sidebar/Sidebar.story.js [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+import Sidebar from './Sidebar';
+
+const stories = storiesOf('Sidebar', module);
+
+const SidebarExample = (props) => {
+  return (
+    <div className="flex">
+      <div className="flex--1 soft">
+        <h2>Content</h2>
+        <p className="soft">
```
Comment (daverau-optimizely) [inline src/components/Sidebar/Sidebar.story.js]: remove inner `soft`

File: src/components/Sidebar/tests/index.js [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+      expect(childComponent.text()).toEqual(mockSidebarContent);
+    });
+
+    it('should add docked class when Sidebar is docked', function() {
+      expect(component.find('[data-test-section="oui-sidebar"]').hasClass('oui-sidebar--docked')).toEqual(false);
+      component.setProps({ docked: true });
+      expect(component.find('[data-test-section="oui-sidebar"]').hasClass('oui-sidebar--docked')).toEqual(true);
+    });
+
+    it('should add open left class when Sidebar anchor is left', function() {
```
Comment (daverau-optimizely) [inline src/components/Sidebar/tests/index.js]: Personally I never know how deep/extensive to test components like this when we're just testing for the presence of CSS classes. 

@pauloptimizely Do you feel like jest snapshots covers this test case or should we be more explicit on testing CSS classes like you've done here?

**Comment by jamesopti** [comment]
> LGTM, is this a copy of the sidebar from the monolith react_components? If so, I want to make sure there is a ticket to migrate the monolith to migrate to using this OUI component. @daverau-optimizely you may want to review the CSS as I'm not familiar enough with OUI to review that properly.

+100 to this - Lets ensure this component can be dropped into the monolith usage.

yay for consistency and reusage!

File: src/components/Sidebar/index.scss (hunk 1 above)
Comment (pauloptimizely) [reply src/components/Sidebar/index.scss]: good call!! I totally agree we should use existing classes 👍

File: src/components/Sidebar/tests/index.js (hunk 5 above)
Comment (pauloptimizely) [reply src/components/Sidebar/tests/index.js]: For this test case, explicit testing allows for more code flexibility. For example, let's say we decided to add a new rule, that happens to add a new class when the Sidebar is open left. Having an explicit test suite will allow a developer to implement the new rule without having to worry about break unrelated tests. If we use snapshot testing, we'll have to constantly keep track of changes to the snapshot when code details change

**Comment by pauloptimizely** [comment]
@jamesopti @quarklemotion I take a look at making this component more compatibly with the Sidebar found to react_components directory, the current implement of this component is very loosely based on monolith's sidebar. The main difference will be, the introduction of "docking". Docked sidebars are fixed and non-docked sidebars will act more like lego drawers.

**Comment by pauloptimizely** [comment]
@daverau-optimizely I just saw your question about the event debugger. I'm creating this component for the comments sidebar in Program Management. The comments section doesn't need a close button, but I can added that feature in if want, let me know

**Comment by pauloptimizely** [comment]
@jamesopti @quarklemotion I tested this component out in the monolith and it will work as a drop-in replacement to the current sidebar found in react_components. The only caveat is you'll have add the correct props

```js
<Sidebar
  isOpen={ true }
  docked={ true }
  border={ true }
  anchor='left'
  maxWidth={ 300 }
/>
```

@daverau-optimizely I ran into some issues building storybook. I'm going to make a separate PR, as soon as find a solution

```
$ npm run build-storybook

> optimizely-oui@40.4.1 build-storybook /Users/pserraino/dev/oui
> build-storybook -c .storybook -o ./storybook

info @storybook/react v3.4.8
info
info => Loading custom .babelrc
info => Loading custom addons config.
info => Loading custom webpack config (extending mode).
info Building storybook ...
(node:80649) DeprecationWarning: loaderUtils.parseQuery() received a non-string value which can be problematic, see https://github.com/webpack/loader-utils/issues/56
parseQuery() will be replaced with getOptions() in the next major version of loader-utils.
[BABEL] Note: The code generator has deoptimised the styling of "/Users/pserraino/dev/oui/node_modules/lodash/lodash.js" as it exceeds the max of "500KB".
[BABEL] Note: The code generator has deoptimised the styling of "/Users/pserraino/dev/oui/node_modules/lodash/lodash.js" as it exceeds the max of "500KB".
ERR! Failed to build the storybook
ERR! ./node_modules/react-split-pane/dist/index.esm.js
ERR! Module build failed: Error: Couldn't find preset "./.babelrc.js" relative to directory "/Users/pserraino/dev/oui/node_modules/react-split-pane"
ERR!     at /Users/pserraino/dev/oui/node_modules/babel-core/lib/transformation/file/options/option-manager.js:293:19
ERR!     at Array.map (<anonymous>)
ERR!     at OptionManager.resolvePresets (/Users/pserraino/dev/oui/node_modules/babel-core/lib/transformation/file/options/option-manager.js:275:20)
ERR!     at OptionManager.mergePresets (/Users/pserraino/dev/oui/node_modules/babel-core/lib/transformation/file/options/option-manager.js:264:10)
ERR!     at OptionManager.mergeOptions (/Users/pserraino/dev/oui/node_modules/babel-core/lib/transformation/file/options/option-manager.js:249:14)
ERR!     at OptionManager.init (/Users/pserraino/dev/oui/node_modules/babel-core/lib/transformation/file/options/option-manager.js:368:12)
ERR!     at File.initOptions (/Users/pserraino/dev/oui/node_modules/babel-core/lib/transformation/file/index.js:212:65)
ERR!     at new File (/Users/pserraino/dev/oui/node_modules/babel-core/lib/transformation/file/index.js:135:24)
ERR!     at Pipeline.transform (/Users/pserraino/dev/oui/node_modules/babel-core/lib/transformation/pipeline.js:46:16)
ERR!     at transpile (/Users/pserraino/dev/oui/node_modules/babel-loader/lib/index.js:46:20)
ERR!     at Object.module.exports (/Users/pserraino/dev/oui/node_modules/babel-loader/lib/index.js:163:20)
ERR!  @ ./node_modules/@storybook/ui/dist/modules/ui/components/layout/index.js 41:22-49
ERR!  @ ./node_modules/@storybook/ui/dist/modules/ui/containers/layout.js
ERR!  @ ./node_modules/@storybook/ui/dist/modules/ui/routes.js
ERR!  @ ./node_modules/@storybook/ui/dist/modules/ui/index.js
ERR!  @ ./node_modules/@storybook/ui/dist/index.js
ERR!  @ ./node_modules/@storybook/core/dist/client/manager/index.js
ERR!  @ multi ./node_modules/@storybook/react/dist/server/config/polyfills.js ./.storybook/addons.js ./node_modules/@storybook/core/dist/client/manager/index.js
info Building storybook completed.
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! optimizely-oui@40.4.1 build-storybook: `build-storybook -c .storybook -o ./storybook`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the optimizely-oui@40.4.1 build-storybook script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/pserraino/.npm/_logs/2018-10-03T19_41_04_300Z-debug.log
```

**Comment by daverau-optimizely** [comment]
@pauloptimizely No need to add the close button for this case right now.

I would like to see those minor CSS fixes, and otherwise lgtm.

File: src/components/Sidebar/index.scss [hunk 6]
```diff
@@ -0,0 +1,5 @@
+.oui-sidebar {
+  transition: all 0.3s;
+  top: 0;
+  z-index: 99;
```
[the remark this answers is not in the document]
Comment (pauloptimizely) [reply src/components/Sidebar/index.scss]: Good call!! 💯  I was wondering if we had z-index values! I think `z-index 99` will work

**Pull request merged**